### DISCUSSION TITLE
How to use CopyMessages?

### DISCUSSION THREAD

**Discussion by cnekmp** [mymmrac/telego#204] [Q&A] [answered]
I'm developing a simple echo bot, that would forward (post) received message to the channel.

It's easy to do when the message contains single media with caption. While when the user sends multiple media in a single message, bot will forward every media as a new message. I've tried to use bot.CopyMessages but it fails as message_ids are null. How can I get those values?

Code:

```
func main() {
	bot, err := telego.NewBot(botToken, telego.WithDefaultDebugLogger())
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	botUser, err := bot.GetMe()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	fmt.Printf("Bot user: %+v\n", botUser)

	updates, _ := bot.UpdatesViaLongPolling(nil)

	bh, _ := th.NewBotHandler(bot, updates)

	bh.HandleMessage(func(bot *telego.Bot, message telego.Message) {

		myStruct := &telego.CopyMessagesParams{}
		chatID := tu.ID(message.Chat.ID)

		b := myStruct.WithFromChatID(chatID)

		_, _ = bot.CopyMessages(b)
	})

	defer bh.Stop()
	defer bot.StopLongPolling()
	bh.Start()
}
```

![image](https://github.com/user-attachments/assets/1ed1a4b2-28b8-46fd-b0b2-9c90cc5b3a59)

**Comment by mymmrac** [comment]
Hi, please take a look at `CopyMessageParams`

```go
// CopyMessageParams - Represents parameters of copyMessage method.
type CopyMessageParams struct {
	// ChatID - Unique identifier for the target chat or username of the target channel (in the format
	// @channel_username)
	ChatID ChatID `json:"chat_id"`

	// MessageThreadID - Optional. Unique identifier for the target message thread (topic) of the forum; for
	// forum supergroups only
	MessageThreadID int `json:"message_thread_id,omitempty"`

	// FromChatID - Unique identifier for the chat where the original message was sent (or channel username in
	// the format @channel_username)
	FromChatID ChatID `json:"from_chat_id"`

	// MessageID - Message identifier in the chat specified in from_chat_id
	MessageID int `json:"message_id"`

	// ...
}
```

It has 3 required parameters: `ChatID`, `FromChatID` and `MessageID`, you only provided `ChatID`, provide two other and it should work

**Reply by mymmrac** [reply]
Also, use `CopyMessage` instead of `CopyMessages`, in your case you are copying only 1 message

**Reply by cnekmp** [reply]
> Hi, please take a look at `CopyMessageParams`
> 
> ```go
> // CopyMessageParams - Represents parameters of copyMessage method.
> type CopyMessageParams struct {
> 	// ChatID - Unique identifier for the target chat or username of the target channel (in the format
> 	// @channel_username)
> 	ChatID ChatID `json:"chat_id"`
> 
> 	// MessageThreadID - Optional. Unique identifier for the target message thread (topic) of the forum; for
> 	// forum supergroups only
> 	MessageThreadID int `json:"message_thread_id,omitempty"`
> 
> 	// FromChatID - Unique identifier for the chat where the original message was sent (or channel username in
> 	// the format @channel_username)
> 	FromChatID ChatID `json:"from_chat_id"`
> 
> 	// MessageID - Message identifier in the chat specified in from_chat_id
> 	MessageID int `json:"message_id"`
> 
> 	// ...
> }
> ```
> 
> It has 3 required parameters: `ChatID`, `FromChatID` and `MessageID`, you only provided `ChatID`, provide two other and it should work

No, I need to use `bot.CopyMessages` instead of `bot.CopyMessage`. As you can see, bot separates messages into two for each attached image instead of sending them as one.
But the issue is, how can I fill this struct? Because `CopyMessagesParams` struct requires `MessageIDs`. How can I get and parse them?

And second question arise: How should I know if the message contains only one image or multiple ones so I could use `CopyMessage` and `CopyMessages` respectively?
![image](https://github.com/user-attachments/assets/b643c178-ea43-48e0-8005-5a291d571c0a)

Code I've used in provided example:

func main() {

	bot, err := telego.NewBot(botToken, telego.WithDefaultDebugLogger())
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	updates, _ := bot.UpdatesViaLongPolling(nil)

	bh, _ := th.NewBotHandler(bot, updates)

		bh.HandleMessage(func(bot *telego.Bot, message telego.Message) {
		chatID := tu.ID(message.Chat.ID)
		_, _ = bot.CopyMessage(
			tu.CopyMessage(chatID, chatID, message.MessageID),
		)
	})
	})

	defer bh.Stop()
	defer bot.StopLongPolling()

	bh.Start()

}

**Reply by mymmrac** [reply]
You can rely only on what Telegram provides, so you need to use `message.MessageID`, I don't know the reason why Telegram sends it as separate messages since it only provides one ID

**Comment by cnekmp** [comment]
Ok, I've tried to collect MediaGroup but when I send "yes" to provide it back to user I get an error:
```
DEBUG API response sendMediaGroup: Ok: false, Err: [400 "Bad Request: can't parse InputMedia: type \"\" is unsupported"]
panic: telego: sendMediaGroup(): api: 400 "Bad Request: can't parse InputMedia: type \"\" is unsupported"

goroutine 1 [running]:
main.main()
        /home/user/forwardbot/forwardbot.go:44 +0x5b4
exit status 2
```

Here is the code:

```
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/mymmrac/telego"
)

const (
	botToken  = "TOKEN HERE"
)

func main() {

	bot, err := telego.NewBot(botToken, telego.WithDefaultDebugLogger())
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	media := []telego.InputMedia{}
	updates, _ := bot.UpdatesViaLongPolling(nil)

	defer bot.StopLongPolling()

	for update := range updates {
		if update.Message != nil {
			if update.Message.MediaGroupID != "" {
				switch {
				case update.Message.Video != nil:
					media = append(media, &telego.InputMediaVideo{Media: telego.InputFile{FileID: update.Message.Video.FileID}})
				case update.Message.Photo != nil:
					media = append(media, &telego.InputMediaPhoto{Media: telego.InputFile{FileID: update.Message.Photo[len(update.Message.Photo)-1].FileID}})
				default:
					log.Println("not implemented")
				}
			}
			// Send collected MediaGroup if "yes" message has been received
			if update.Message.Text == "yes" {
				_, err = bot.SendMediaGroup(&telego.SendMediaGroupParams{ChatID: update.Message.Chat.ChatID(), Media: media})
				if err != nil {
					panic(err)
				}
			}
		}
	}
}
```

**Comment by cnekmp** [comment, marked as answer]
Ok, Found a solution. I should provide Type: "video" or Type: "photo" in struct. Providing working code. Closing discussion

```
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/mymmrac/telego"
)

const (
	botToken = "<TOKEN>"
)

func main() {

	bot, err := telego.NewBot(botToken, telego.WithDefaultDebugLogger())
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	media := []telego.InputMedia{}
	updates, _ := bot.UpdatesViaLongPolling(nil)

	defer bot.StopLongPolling()

	for update := range updates {
		if update.Message != nil {
			if update.Message.MediaGroupID != "" {
				switch {
				case update.Message.Video != nil:
					media = append(media, &telego.InputMediaVideo{Type: "video", Media: telego.InputFile{FileID: update.Message.Video.FileID}})
				case update.Message.Photo != nil:
					media = append(media, &telego.InputMediaPhoto{Type: "photo", Media: telego.InputFile{FileID: update.Message.Photo[len(update.Message.Photo)-1].FileID}})
				default:
					log.Println("not implemented")
				}
			}
			// Send collected MediaGroup if "yes" message has been received
			if update.Message.Text == "yes" {
				_, err = bot.SendMediaGroup(&telego.SendMediaGroupParams{ChatID: update.Message.Chat.ChatID(), Media: media})
				if err != nil {
					panic(err)
				}
			}
		}
	}
}

```
